crypto/provider.c

```diff
@@ -7,28 +7,40 @@
  * https://www.openssl.org/source/license.html
  */
 
+#include <string.h>
 #include <openssl/err.h>
 #include <openssl/cryptoerr.h>
 #include <openssl/provider.h>
 #include <openssl/core_names.h>
 #include "internal/provider.h"
+#include "provider_local.h"
 
 OSSL_PROVIDER *OSSL_PROVIDER_try_load(OSSL_LIB_CTX *libctx, const char *name,
                                       int retain_fallbacks)
 {
-    OSSL_PROVIDER *prov = NULL;
+    OSSL_PROVIDER *prov = NULL, *actual;
+    int isnew = 0;
 
     /* Find it or create it */
-    if ((prov = ossl_provider_find(libctx, name, 0)) == NULL
-        && (prov = ossl_provider_new(libctx, name, NULL, 0)) == NULL)
+    if ((prov = ossl_provider_find(libctx, name, 0)) == NULL) {
+        if ((prov = ossl_provider_new(libctx, name, NULL, 0)) == NULL)
```

Comment:
Isn't this going to be a race between two threads creating the provider after not finding it?  The second will error and fail even though the provider exists...

This hasn't changed so no need to fix it here but perhaps raise an issue if there is a problem.

Reply:
No, I don't think there is a problem here. The two threads will race and both will create the provide and activate it. However, when they come to add it to the store the first thread will "win" and succeed. The second thread will call `ossl_provider_add_to_store`. That function considers it a *success* if the provider already exists. It deactivates and frees the provider that was attempted to be added, and returns the provider that already exists in the store in the `actual` variable below.

So in this way we will initialise two instances of the provider, but only one will actually get used, and the other will be cleanly closed down and freed.